### PR TITLE
Add batch import tool to DB import

### PR DESCRIPTION
**Pull request by beckerhe**
This is adding a new tool to the CLI, the batch import.

The batch import can scan through all the files in an existing GCS bucket and trigger the import pipeline as if all those files were just uploaded. It's usually needed when a new pipeline is set up and when we want to import all the existing data.

### CODE REVIEW

File: devtools/db_import/db_import/batch_import.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        db_client.insert_rows(table, rows)
+      print("Done.")
+    except NoRuleAppliesError:
+      # This happens so often, that we don't wanna clutter the output, so
+      # we rather do a carriage return and re-use the same line.
+      print("No rule applies.\r" + " " * 250, end="\r")
+    except BenchmarkRunAlreadyPresentError:
+      print("Data already present.")
+    except Exception:
+      print("Failure.")
```
Comment (pzread) [inline devtools/db_import/db_import/batch_import.py:93]: Maybe dump the exception?

File: devtools/db_import/db_import/db.py [hunk 2]
```diff
@@ -47,3 +47,38 @@ def insert_rows(self, table: Union[str, Table], rows: Iterable[Mapping[str,
...
+    parameters: dict[str, str],
+) -> bool:
+  job_config = bigquery.QueryJobConfig(query_parameters=[
+      bigquery.ScalarQueryParameter(key, "STRING", value)
+      for key, value in parameters.items()
+  ])
+
+  rows = list(
+      client.query(
+          f"SELECT COUNT(*) as count FROM ({sql})",
```
Comment (pzread) [inline devtools/db_import/db_import/db.py:64]: Will this result in SQL injection? How do we ensure the `sql` is trustable

Comment (beckerhe) [reply devtools/db_import/db_import/db.py:64]: We ensure that `sql` is trustable the same way we ensure that the sql argument to `client.query` is trustable.

I see this function as a convenience wrapper around `client.query` - which also takes an arbitrary SQL string as input, hence there is no bigger risk than using `client.query` directly, right?

I could potentially remove the count aggregation here and make it the caller's responsibility to return a `cound` field right away. Then I could just forward the sql untouched. Would that make more sense to you?

File: devtools/db_import/db_import/batch_import.py (hunk 1 above)
Comment (beckerhe) [reply devtools/db_import/db_import/batch_import.py]: Good idea. Done.

File: devtools/db_import/db_import/db.py (hunk 2 above)
Comment (beckerhe) [reply devtools/db_import/db_import/db.py:64]: Another option is to hand off this task to a SQL parser library like sqlglot.

The following expression should do the same:

```python
sqlglot.subquery(sqlglot.parse_one(sql)).select("COUNT(*) as count").sql()
```

sqlglot parses the subquery into an AST and reemits it in the `.sql()` call. This should result in something that fails whenever `sql` is not a valid select expression (modulo sqlglot bugs of course)

WDYT about that?

Comment (pzread) [reply devtools/db_import/db_import/db.py:64]: Makes sense. I also see `bigquery.QueryJobConfig` is used to pass parameters from untrusted sources. The current approach is reasonable to me.

**Review by pzread — APPROVED** [review]

File: devtools/db_import/db_import/batch_import.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import argparse
+import sys
+
+from google.cloud import bigquery
+from google.cloud import storage
+from typing import Optional, Any, Dict
+
+from db_import.process import process_single_file
+from db_import.process import NoRuleAppliesError
+from db_import.rules import BenchmarkRunAlreadyPresentError
```
[the remark this answers is not in the document]
Comment (beckerhe) [reply devtools/db_import/db_import/batch_import.py]: Yes. Thanks for pointing that out.

File: devtools/db_import/db_import/db.py [hunk 4]
```diff
@@ -47,3 +47,38 @@ def insert_rows(self, table: Union[str, Table], rows: Iterable[Mapping[str,
 
   def get_table(self, table: Union[str, Table]) -> Table:
     ...
+
+
+def query_returns_non_empty_result(
+    client: Client,
+    sql: str,
+    parameters: dict[str, str],
```
[the remark this answers is not in the document]
Comment (beckerhe) [reply devtools/db_import/db_import/db.py]: Oh, yes, yes. Missed again. Thx.

File: devtools/db_import/db_import/db.py (hunk 2 above)
Comment (beckerhe) [reply devtools/db_import/db_import/db.py:64]: SG. Leving as-is.

**Pull request merged**